TombEngine/Game/Lara/lara_one_gun.cpp

```diff
@@ -556,109 +561,98 @@ void UndrawShotgunMeshes(ItemInfo& laraItem, LaraWeaponType weaponType)
+	harpoonItem.Model.Color = Vector4(0.5f, 0.5f, 0.5f, 1.0f);
 
 	if (!ammo.HasInfinite())
 		ammo--;
 
-	auto& item = g_Level.Items[itemNumber];
-
-	item.Model.Color = Vector4(0.5f, 0.5f, 0.5f, 1.0f);
-	item.ObjectNumber = ID_HARPOON;
-	item.RoomNumber = laraItem.RoomNumber;
+	if (pose.has_value())
+	{
+		harpoonItem.Pose.Position = pose->Position;
+		harpoonItem.RoomNumber = laraItem.RoomNumber;
 
-	auto jointPos = GetJointPosition(&laraItem, LM_RHAND, Vector3i(-2, 373, 77));
-	int floorHeight = GetCollision(jointPos.x, jointPos.y, jointPos.z, item.RoomNumber).Position.Floor;
+		InitializeItem(itemNumber);
 
-	if (floorHeight >= jointPos.y)
-	{
-		item.Pose.Position = jointPos;
+		harpoonItem.Pose.Orientation = EulerAngles(
+			player.LeftArm.Orientation.x + laraItem.Pose.Orientation.x,
+			player.LeftArm.Orientation.y + laraItem.Pose.Orientation.y,
+			0);
 	}
 	else
 	{
-		item.Pose.Position = Vector3i(laraItem.Pose.Position.x, jointPos.y, laraItem.Pose.Position.z);
-		item.RoomNumber = laraItem.RoomNumber;
-	}
+		auto jointPos = GetJointPosition(&laraItem, LM_RHAND, Vector3i(-2, 373, 77));
+		harpoonItem.RoomNumber = laraItem.RoomNumber;
 
-	InitializeItem(itemNumber);
+		int floorHeight = GetCollision(jointPos.x, jointPos.y, jointPos.z, harpoonItem.RoomNumber).Position.Floor;
+		if (floorHeight >= jointPos.y)
+		{
+			harpoonItem.Pose.Position = jointPos;
+		}
+		else
+		{
+			harpoonItem.Pose.Position = Vector3i(laraItem.Pose.Position.x, jointPos.y, laraItem.Pose.Position.z);
+			harpoonItem.RoomNumber = laraItem.RoomNumber;
+		}
 
-	item.Pose.Orientation = EulerAngles(
-		player.LeftArm.Orientation.x + laraItem.Pose.Orientation.x,
-		player.LeftArm.Orientation.y + laraItem.Pose.Orientation.y,
-		0);
+		InitializeItem(itemNumber);
 
-	if (!player.LeftArm.Locked)
-		item.Pose.Orientation += player.ExtraTorsoRot;
+		harpoonItem.Pose.Orientation.x = player.LeftArm.Orientation.x + laraItem.Pose.Orientation.x;
+		harpoonItem.Pose.Orientation.z = 0;
+		harpoonItem.Pose.Orientation.y = player.LeftArm.Orientation.y + laraItem.Pose.Orientation.y;
 
-	item.Pose.Orientation.z = 0;
-	item.Animation.Velocity.z = HARPOON_VELOCITY * phd_cos(item.Pose.Orientation.x);
-	item.Animation.Velocity.y = -HARPOON_VELOCITY * phd_sin(item.Pose.Orientation.x);
-	item.HitPoints = HARPOON_TIME;
+		if (!player.LeftArm.Locked)
+			harpoonItem.Pose.Orientation += player.ExtraTorsoRot;
+	}
+
+	harpoonItem.Animation.Velocity.z = HARPOON_VELOCITY;
+	harpoonItem.HitPoints = HARPOON_TIME;
 
-	Rumble(0.2f, 0.1f);
 	AddActiveItem(itemNumber);
 
+	harpoonItem.ItemFlags[0] = (int)ProjectileType::Normal;
+
+	Rumble(0.2f, 0.1f);
+
 	Statistics.Level.AmmoUsed++;
 	Statistics.Game.AmmoUsed++;
-
-	return &item;
 }
-
+	
 void HarpoonBoltControl(short itemNumber)
 {
 	auto& harpoonItem = g_Level.Items[itemNumber];
 
-	if (harpoonItem.HitPoints < HARPOON_TIME)
+	if (TestEnvironment(ENV_FLAG_WATER, &harpoonItem))
 	{
-		if (harpoonItem.HitPoints > 0)
```

Comment:
Where this code have gone? It shattered harpoon after a time interval when it stucked into room geometry!